### PR TITLE
Enhancement/support updated execution config json cli execution engine

### PR DESCRIPTION
**Pull request by vyasprateek**
- Ginger CLI execution engine Enhancements, to support extended configuration JSON (Virtual BF)
- Support for existing Business Flows in solution but not part of Runner
- Virtual BF with Shared Repository Activities support
- Unit tests added to cover the updated scenarios
- 
- Pending - Testing CLI JSON Fields in Runset Shortcut

### CODE REVIEW

File: Ginger/GingerCoreNET/RunLib/CLILib/CLIDynamicFile.cs [hunk 1]
```diff
@@ -217,7 +217,7 @@ private void LoadALMDetailsFromJSON(GingerExecConfig gingerExecConfig)
             foreach (AlmDetails almDetails in gingerExecConfig.AlmsDetails)
             {
                 ALMIntegrationEnums.eALMType almTypeToConfigure;
-                if (Enum.TryParse<ALMIntegrationEnums.eALMType>(almDetails.ALMType, out almTypeToConfigure))
+                if (Enum.TryParse(almDetails.ALMType.ToString(), out almTypeToConfigure))
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/RunLib/CLILib/CLIDynamicFile.cs:220]: why this change was needed

File: Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs [hunk 2]
```diff
@@ -750,6 +753,10 @@ private static GingerExecConfig GetGingerExecConfigurationObject(Solution soluti
                     jsonReportConfig.Active = runSetOperation.Active;
                     runset.Operations.Add(jsonReportConfig);
                 }
+                else if (runSetOperation is RunSetActionAutomatedALMDefects)
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs:756]: RunSetActionAutomatedALMDefects is not the right operation we adding support for from CLI JSON, only Export Results to ALM operation (not defects)

File: Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs [hunk 3]
```diff
@@ -932,13 +961,68 @@ public static void CreateUpdateRunSetFromJSON(RunsetExecutor runsetExecutor, Gin
...
+                                    /// Fetch BF & use the same for execution
+                                    if (businessFlowConfig.ID != null || !string.IsNullOrEmpty(businessFlowConfig.Name))
+                                    {
+                                        bf = (BusinessFlow)FindItemByIDAndName<BusinessFlow>(
+                                         new Tuple<string, Guid?>(nameof(BusinessFlow.Guid), businessFlowConfig.ID),
+                                         new Tuple<string, string>(nameof(BusinessFlow.Name), businessFlowConfig.Name),
+                                         WorkSpace.Instance.SolutionRepository.GetAllRepositoryItems<BusinessFlow>());
+                                    }
+                                }
+                                catch(Exception exc)
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs:982]: no need this try-catch, if ID was not found so exception needs to be thrown and execution to fail like we having for any other entity ID in the JSON

File: Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs [hunk 4]
```diff
@@ -1038,7 +1122,7 @@ public static void CreateUpdateRunSetFromJSON(RunsetExecutor runsetExecutor, Gin
                                 }
                             }
 
-                            if (!dynamicRunsetConfigs.Exist)
+                            if (!dynamicRunsetConfigs.Exist || (businessFlowConfig.Exist.HasValue && !businessFlowConfig.Exist.Value))
                             {
                                 gingerRunner.BusinessFlowsRunList.Add(businessFlowRun);
                             }
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs:1128]: need to add also for line 1130 the check if Exist do appear in the JSON for Runner with value False

File: Ginger/GingerCoreNET/RunLib/CLILib/CLIDynamicFile.cs (hunk 1 above)
Comment (vyasprateek) [reply Ginger/GingerCoreNET/RunLib/CLILib/CLIDynamicFile.cs:220]: Because almDetails.ALMType property's type is changed from string to enum

File: Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs (hunk 3 above)
Comment (vyasprateek) [reply Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs:982]: This is for creating Virtual Business Flows which doesn't exists in the solution

Comment (MeniKadosh1) [reply Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs]: we agreed that no need to create virtual if ID provided but not found, need to throw exception

Comment (vyasprateek) [reply Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs]: This is outdated changes Meni. In the latest commit, it is fixed @ https://github.com/Ginger-Automation/Ginger/pull/2283/commits/766ecd5f48abad5a24b5fc531fd39eee4c68bc1d

image

**Review by MeniKadosh1 — APPROVED** [review]

**Review by MeniKadosh1 — APPROVED** [review]

**Review by MeniKadosh1 — APPROVED** [review]
